src/zarr/storage/_cache.py

```diff
@@ -0,0 +1,735 @@
+import io
+import warnings
+from collections import OrderedDict
+from collections.abc import AsyncIterator, Iterable, Iterator
+from pathlib import Path
+from threading import Lock
+from typing import Any, TypeAlias
+
+import numpy as np
+
+from zarr.abc.store import OffsetByteRequest, RangeByteRequest, Store, SuffixByteRequest
+from zarr.core.buffer import Buffer, BufferPrototype
+from zarr.core.buffer.core import default_buffer_prototype
+from zarr.storage._utils import normalize_path
+
+ByteRequest: TypeAlias = RangeByteRequest | OffsetByteRequest | SuffixByteRequest
+
+
+def buffer_size(v: Any) -> int:
+    """Calculate the size in bytes of a value, handling Buffer objects properly."""
+    if hasattr(v, "__len__") and hasattr(v, "nbytes"):
+        # This is likely a Buffer object
+        return int(v.nbytes)
+    elif hasattr(v, "to_bytes"):
+        # This is a Buffer object, get its bytes representation
+        return len(v.to_bytes())
+    elif isinstance(v, (bytes, bytearray, memoryview)):
+        return len(v)
+    else:
+        # Fallback to numpy
+        return int(np.asarray(v).nbytes)
+
+
+def _path_to_prefix(path: str | None) -> str:
+    # assume path already normalized
+    if path:
+        prefix = path + "/"
+    else:
+        prefix = ""
+    return prefix
+
+
+def _listdir_from_keys(store: Store, path: str | None = None) -> list[str]:
+    # assume path already normalized
+    prefix = _path_to_prefix(path)
+    children: set[str] = set()
+    # Handle both Store objects and dict-like objects
+    if hasattr(store, "keys") and callable(store.keys):
+        keys = [str(k) for k in store.keys()]  # Ensure keys are strings  # noqa: SIM118
+    else:
+        # For stores that don't have keys method, we can't list them
+        return []
+
+    for key in keys:
+        if key.startswith(prefix) and len(key) > len(prefix):
+            suffix = key[len(prefix) :]
+            child = suffix.split("/")[0]
+            children.add(child)
+    return sorted(children)
+
+
+def listdir(store: Store, path: Path | None = None) -> list[str]:
+    """Obtain a directory listing for the given path. If `store` provides a `listdir`
+    method, this will be called, otherwise will fall back to implementation via the
+    `MutableMapping` interface."""
```

Comment:
```suggestion
    """
    Obtain a directory listing for the given path.
    
    If `store` provides a `listdir`
    method, this will be called, otherwise will fall back to implementation via the
    `MutableMapping` interface.
    """
```
If possible, [the first line of a docstring should be a one line summary.](https://numpydoc.readthedocs.io/en/latest/format.html#short-summary)